### ISSUE TITLE
View Constrained To Parent And Sibling Fails

### ISSUE THREAD

**Issue by lprichar** [lprichar/EasyLayout.Forms#3]
Basically right now this fails:

```
            _relativeLayout.ConstrainLayout(() =>
                _topLabel.Top() == _relativeLayout.Top()
                && _topLabel.Left() == _relativeLayout.Left()
                && _topLabel.Right() == _relativeLayout.Right()
                
                && _bottomLabel.Top() == _topLabel.Bottom()
                && _bottomLabel.Left() == _relativeLayout.Left()
                && _bottomLabel.Right() == _relativeLayout.Right()
                && _bottomLabel.Bottom() == _relativeLayout.Bottom()
```

Because it translates to _bottomLabel neeeding a height constraint like Parent.Height - _topLabel.Height.

**Comment by lprichar** [comment]
I've had an extensive conversation with @appbureauet about this that I'll post in here for context:

@appbureauet 6/21:

 What I am experiencing problems with right now is that RelativeLayout seems to be taller than the screen.

To illustrate my problem I have created two gists. I would like for my code to look like this:

https://gist.github.com/appbureauet/41446d16b92e8b10ccbb1552d65961d2

Look at line 10 where I try to align my ListView’s bottom to the RelativeLayout’s bottom.
This doesn’t work, I am unsure if it is because the bottom operation fails or the height needs to be set explicitly as well.

Currently I have to do this in my code:

https://gist.github.com/appbureauet/41446d16b92e8b10ccbb1552d65961d2/5e6da822e2b2053b30891781fb445532d14517a8

Doing it this way creates the need for branching out the code and knowing the searchBarHeight constants at compile time.

So I am thinking there could be three solutions to this:

1 - I am doing it wrong and you show me the way :)
2 - We should look at improving how the height/width is set when using both top and bottom or left and right.
3 - We should look at how RelativeLayout is positioned if that is the problem.

Hope this makes sense?

@lprichar 6/23

Congratulations I believe you found a bug thus #2 in the winner.  :)  I documented what I believe is the problem here in playground: 
https://github.com/lprichar/EasyLayout.Forms/commit/5b1ef446a9d9f346ded62198fba4771e3d934610

I'm going to take a stab at it now.

@lprichar 6/24

Ugh, this is a hard problem. I'm looking forward to a real RelativeLayout in XF3.  Until then I took a 1st stab at it here: 

https://github.com/lprichar/EasyLayout.Forms/commit/5b1ef446a9d9f346ded62198fba4771e3d934610 

Sadly that only works if there is one prior view that the bottom constrained one is dependent upon, and it has an explicit height.  I'll have to think more about this some more including how to handle the implicit heights like in your example.

Incidentally, the problem is the bottom view's top and bottom constraint translate to a height constraint.  That height constraint is dependent upon both the parent relative layout (Constraint.RelativeToParent()) AND another view (Constraint.RelativeToView()).    The fix there continues to choose the RelativeToParent option but then subtracts out the other view's height if it was specified.

@lprichar 6/24

After sleeping on it I'm wondering if we couldn't use the RelativeToView() option instead, and then calculate the container's height using something like otherview.Parent.Height.

Incidentally two ideas I had that may or may not fit in for this particular issue:
1. Add in a new "Reduce" step at the highest level into ConstrainLayout() such that at a high level it 1. parses into Constraints. 2. Converts constraints into Rules. 3. Reduces the rules to remove/replace bottom/right rules with height/width rules. 4. Adds the reduced rules into relativeLayout.Children.Add calls.  Basically this would simplify UpdateLayoutParamsWithRules() so that it has no logic in it at all.
2. Unit test at the level of 1-3 above.  Step 4, the relativeLayout.Children.Add call should be so simple it doesn't need unit testing.

@appbureauet 6/27

I have tried looking at it, but I am kinda stuck because I can’t figure out how to make it work with more than two views.

I have created a quick thing here that shows how it could be done if you only have two views - 

https://github.com/appbureauet/EasyLayout.Forms/commit/02c57deeadaf66d98e119a5afb00a6135cb0e612

I can’t seem to get my head around how to get the correct height/width when RelativeLayout only allows one height constraint.

Anyway my head is melted cheese right now - maybe a solution will pop up in the next few days. If not maybe you can find a better solution by looking at what I tried 😊

**Comment by lprichar** [comment]
@appbureauet So I took a fresh stab at this in just a very narrow context (just solving the problem documented in PlaygroundPage.cs) and I think I pretty much came up with the same solution as you but perhaps with less code.  That's here: https://github.com/lprichar/EasyLayout.Forms/commit/d39435b6f60bb5db2f4619398ca00949577958f3.  That needs to be more generic of course but does it solve your original problem?  It seems like we might not need your GetDependentView() function if we just return the non-parent constraint in GetRelativeHeightConstraint().  Let me know what you think.

**Comment by appbureauet** [comment]
@lprichar Thanks for the update. I like your new approach!
And it does indeed solve the problem if you have the exact use case where a second view is dependent  on another view that is placed at the top and it needs to fill the remaining space.

But I think we need to dig a bit deeper so that we can solve this in a more generic way.

Currently this does not work if you have a view that is dependent on two other views - where none of them are the parent.

I have create a gist here showing this use case - https://gist.github.com/appbureauet/316837f7f41aac6144cc128f5055283e

I still don't have any bright ideas, but I am thinking about it.
Hopefully we will solve this at some point, I don't know if Xamarin Forms 3 will help?

**Comment by appbureauet** [comment]
I just tried to create another example of the above using native relative layout methods combined with your height calculations. This leeds me to believe that we can solve this!

Something along the lines of:
- If a view depends on other views than the parent then we need to get all those views
- Finally we need to add the views to the layout in the correct order
- The height and width should then be a matter of calling your helper methods on all the dependent views

See lines 42-60 in this example of what I am thinking (in this code all the views are positioned correctly) - https://gist.github.com/appbureauet/cb59cdc5a35e9ace3fe4fbbde9c770c8

**Comment by appbureauet** [comment]
I tried to modify my first attempt to support this way of fixing it and created a pull request.

**Comment by lprichar** [comment]
I think I understand what you'd like to accomplish with this PR.  The gist you provided basically says _middleView.Height should be = _relativeLayout.Height - _topLabel.Height - _bottomLabel.Height, right?  

I'll dig more into the proposed solution later, but first my initial impression is we are trying to solve the wrong problem.  The issue is RelativeLayout just isn't that sophisticated.  Xamarin.Forms RelativeLayout doesn't give us multiple view dependencies because it's a hard problem with significant performance implications.  For instance, what if there's a whole stack of views whose height we need to subtract out, some of whome have explicit heights and others whose heights are intrinsic?  It's like trying to give developers multiple inheritance in a language that only supports single inheritance, it's just never going to work quite right.

I would love to solve that problem, however, I just feel like to do so we'd need to throw out RelativeLayout and start with a custom layout.  In fact that's exactly what I was thinking about when I watched this presentation from Jason Smith: https://youtu.be/sxjOqNZFhKU.  Great talk if you haven't seen it yet.

The thing is I had never intended to re-create iOS's AutoLayout with ELXF once I realized RelativeLayout's deficiencies, because AutoLayout is something Apple spent thousands of man-hours getting perfect.  My original objective was to make the given control better in a code-behind scenario, help throw out XAML, and then to help fix some of it's most common deficiencies, like center aligning things.  

The original example you provided with a bottom-most view that expands out into it's parent and the view above it seems to me to fit into the bucket of a common deficiency.

So given that background, do you still feel like recreating AutoLayout is a problem that ELXF should be attacking and can solve?

**Comment by appbureauet** [comment]
Thanks for the link to Jason Smiths video, it's been idling for too long on my watch later list - I'll watch that soon 👍

Good thoughts about the pros and cons of fighting too much with RelativeLayout. You are right that the new PR could potentially have a negative impact on performance. But if you use it with common sense, eg that you can't have more than one unknown height or width - then I guess it will be fine. Correct me if I am wrong.

> I think I understand what you'd like to accomplish with this PR. The gist you provided basically says _middleView.Height should be = _relativeLayout.Height - _topLabel.Height - _bottomLabel.Height, right?

That is correct. I think that would be nice to be able to do something like this with ELXF.

The PR solves the original problem I had. And it does it in a way where the unconstrained view can sit anyway in the stack. That works for me. But I am very open for suggestions (and I agree - we should not try to recreate AutoLayout completely 😊).

If we go forward with this approach, then it could be nice with an exception to the user if they try to have too many unconstrained views.

**Comment by lprichar** [comment]
Resolved with Pull Request #4
